src/backend/distributed/executor/multi_executor.c

```diff
@@ -235,16 +235,17 @@ CitusExecutorRun(QueryDesc *queryDesc,
 			 * transactions.
 			 */
 			CitusTableCacheFlushInvalidatedEntries();
-
-			/*
-			 * Within a 2PC, when a function is delegated to a remote node, we pin
-			 * the distribution argument as the shard key for all the SQL in the
-			 * function's block. The restriction is imposed to not to access other
-			 * nodes from the current node and violate the transactional integrity
-			 * of the 2PC. Now that the query is ending, reset the shard key to NULL.
-			 */
-			ResetAllowedShardKeyValue();
+			InTopLevelDelegatedFunctionCall = false;
 		}
+
+		/*
+		 * Within a 2PC, when a function is delegated to a remote node, we pin
+		 * the distribution argument as the shard key for all the SQL in the
+		 * function's block. The restriction is imposed to not to access other
+		 * nodes from the current node, and violate the transactional integrity
+		 * of the 2PC. Now that the query is ending, reset the shard key to NULL.
+		 */
+		ResetAllowedShardKeyValue();
```

Comment:
if the "pushed down" function call does multiple things, wouldn't this reset the restriction after the first statement?

Reply:
ah, `ResetAllowedShardKeyValue` changed. Wonder if we can reflect that in the name.

Reply:
Changed to `CheckAndResetAllowedShardKeyValueIfNeeded`()